internal/format/middle_visitor.go

```diff
@@ -0,0 +1,306 @@
+		v.P()
+	}
+}
+
+func (v *middleVisitor) VisitService(element *proto.Service) {
+	v.haveHitNonComment = true
+	v.PComment(element.Comment)
+	if len(element.Elements) == 0 {
+		v.P("service ", element.Name, " {}")
+		v.P()
+		return
+	}
+	v.P("service ", element.Name, " {")
+	v.In()
+	originalParent := v.parent
+	v.parent = element
+	for _, child := range element.Elements {
+		child.Accept(v)
+	}
+	v.parent = originalParent
+	v.Out()
+	v.P("}")
+	v.P()
+}
+
+func (v *middleVisitor) VisitSyntax(element *proto.Syntax) {
+	// done in first pass visitor
+	v.haveHitNonComment = true
+}
+
+func (v *middleVisitor) VisitPackage(element *proto.Package) {
+	// done in first pass visitor
+	v.haveHitNonComment = true
+}
+
+func (v *middleVisitor) VisitOption(element *proto.Option) {
+	v.haveHitNonComment = true
+	// file options done in first pass visitor
+	if v.parent == nil {
+		return
+	}
+	switch v.parent.(type) {
+	case (*proto.Enum):
+		v.POptions(false, element)
+	case (*proto.Message):
+		v.POptions(false, element)
+	case (*proto.Oneof):
+		v.POptions(false, element)
+	case (*proto.Service):
+		v.POptions(false, element)
+	default:
+		v.AddFailure(element.Position, "unhandled child option")
+	}
+}
+
+func (v *middleVisitor) VisitImport(element *proto.Import) {
+	// done in first pass visitor
+	v.haveHitNonComment = true
+}
+
+func (v *middleVisitor) VisitNormalField(element *proto.NormalField) {
+	v.haveHitNonComment = true
+	prefix := ""
+	if element.Repeated {
+		prefix = "repeated "
+	}
+	if v.isProto2 {
+		// technically these are only set if the file is proto2
+		// but doing this just to make srue
+		if element.Required {
+			prefix = "required "
+		} else {
+			prefix = "optional "
+		}
+	}
+	v.PField(prefix, element.Type, element.Field)
+}
+
+func (v *middleVisitor) VisitEnumField(element *proto.EnumField) {
+	v.haveHitNonComment = true
+	v.PComment(element.Comment)
+	if element.ValueOption == nil {
+		v.PWithInlineComment(element.InlineComment, element.Name, " = ", element.Integer, ";")
+		return
+	}
+	v.P(" ", element.Name, " = ", element.Integer, " [")
+	v.In()
+	v.POptions(true, element.ValueOption)
+	v.Out()
+	v.PWithInlineComment(element.InlineComment, "];")
+}
+
+func (v *middleVisitor) VisitEnum(element *proto.Enum) {
+	v.haveHitNonComment = true
+	v.PComment(element.Comment)
+	if len(element.Elements) == 0 {
+		v.P("enum ", element.Name, " {}")
+		v.P()
+		return
+	}
+	v.P("enum ", element.Name, " {")
+	v.In()
+	originalParent := v.parent
+	v.parent = element
+	for _, child := range element.Elements {
+		child.Accept(v)
+	}
+	v.parent = originalParent
+	v.Out()
+	v.P("}")
+	if v.parent == nil {
+		v.P()
+	}
+}
+
+func (v *middleVisitor) VisitComment(element *proto.Comment) {
+	if v.haveHitNonComment {
+		v.PComment(element)
+		v.P()
+	}
+}
+
+func (v *middleVisitor) VisitOneof(element *proto.Oneof) {
+	v.haveHitNonComment = true
+	v.PComment(element.Comment)
+	if len(element.Elements) == 0 {
+		// TODO: is this even legal?
```

Comment:
We should figure this out and see if it's necessary after all.

Reply:
Yea we should, lol.

Reply:
Just checked with the `go`, `gogo`, `python`, `java` protoc plugin, this **is not** allowed: 
```
foo.proto:12:19: Expected type name.
```

Although, it may be worth it to keep this line as that behavior seems reasonable / not surprising, if other plugins allow this?